common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java

```diff
@@ -53,22 +63,130 @@ public AuthorizationRequest getRequest() {
+                                final String failingUrl) {
+        super.onReceivedError(view, errorCode, description, failingUrl);
+
+        // Create result intent when webView received an error.
+        final Intent resultIntent = new Intent();
+        resultIntent.putExtra(AuthenticationConstants.Browser.RESPONSE_ERROR_CODE,
+                "Error Code:" + errorCode);
+        resultIntent.putExtra(AuthenticationConstants.Browser.RESPONSE_ERROR_MESSAGE,
+                description);
+        resultIntent.putExtra(AuthenticationConstants.Browser.RESPONSE_REQUEST_INFO,
+                mRequest);
+
+        // Send the result back to the calling activity
+        mCompletionCallback.onChallengeResponseReceived(AuthenticationConstants.UIResponse.BROWSER_CODE_ERROR, resultIntent);
+    }
+
+    @Override
+    public void onReceivedSslError(final WebView view,
+                                   final SslErrorHandler handler,
+                                   final SslError error) {
+        // Developer does not have option to control this for now
+        super.onReceivedSslError(view, handler, error);
+        handler.cancel();
+
+        // WebView received the ssl error and create the result intent.
+        final Intent resultIntent = new Intent();
+        resultIntent.putExtra(AuthenticationConstants.Browser.RESPONSE_ERROR_CODE,
+                "Code:" + ERROR_FAILED_SSL_HANDSHAKE);
+        resultIntent.putExtra(AuthenticationConstants.Browser.RESPONSE_ERROR_MESSAGE,
+                error.toString());
+        resultIntent.putExtra(AuthenticationConstants.Browser.RESPONSE_REQUEST_INFO,
+                mRequest);
+
+        // Send the result back to the calling activity
+        mCompletionCallback.onChallengeResponseReceived(AuthenticationConstants.UIResponse.BROWSER_CODE_ERROR, resultIntent);
+    }
+
+    @Override
+    public void onPageFinished(final WebView view,
+                               final String url) {
+        super.onPageFinished(view, url);
+        // Once web view is fully loaded,set to visible
+        view.setVisibility(View.VISIBLE);
+    }
+
+    @Override
+    public void onPageStarted(final WebView view,
+                              final String url,
+                              final Bitmap favicon) {
+        checkStartUrl(url);
+        super.onPageStarted(view, url, favicon);
+    }
+
+    private void checkStartUrl(final String url) {
+        //TODO should it return error response when it finds the start url is invalid?
+
+        if (StringUtil.isEmpty(url)) {
+            Logger.verbose(TAG, "onPageStarted: Null url for page to load.");
+            return;
         }
 
-        mContext = context;
-        mRequest = request;
+        final Uri uri = Uri.parse(url);
+        if (uri.isOpaque()) {
+            Logger.verbose(TAG, "onPageStarted: Non-hierarchical loading uri.");
+            Logger.verbosePII(TAG, "start url: " + url);
+            return;

```

Comment:
after reading the context, I think making L178-end of the method in a 'if- else if - else' statement would be more readable. A 'return' statement here hurts the readability .